PCIDSK SDK merge

pciphelippeneveu opened the pull request:
This is the second part of a pull request done a few weeks ago where I only included the cosmetic changes (spaces at end-of-lines, copyright address).

This pull request is the result of a merge of both GDAL's and PCI's changes to the PCIDSK SDK source code over the last 10 years.

I spent a lot of time on minimizing the differences at your end. I carefully merged most of the changes done from GDAL. Even changes to replace usage of EQUALN() and strncmp() with new STARTS_WITH*() macros. I had to write a Python script to merge these changes. There were also a bunch of cosmetic changes like converting tabs into spaces, replacement of NULL with nullptr, usage of override keyword, etc.

Most of the changes are in the ctiledchannel.cpp file. The System Block Map was refactored into 2 different implementations TILEV1 (ASCII) and TILEV2 (Binary) found in a new sdk/blockdir sub-directory. More information was added to the version 1 directory to support lazzy instantiation of the tile layers. By default we now create TILEV2 files, the old ASCII version can still be created using the TILEV1 file creation option.

Support for more data types was added which required changes to cpcidskfile.cpp and cpcidskcreate.cpp. I might have skipped/forgot some GDAL changes in these files.

Quite a lot of changes were done to the cpcidsktoutinmodel.cpp and cpcidskvectorsegment.cpp files. I might have skipped/forgot some GDAL changes in these files.

## Environment

OS: CentOS 8
Compiler: GCC 8.3

rouault commented on gdal/frmts/pcidsk/sdk/blockdir/asciitiledir.cpp:
```diff
@@ -225,7 +225,7 @@ AsciiTileDir::AsciiTileDir(BlockFile * poFile, uint16 nSegment)
  */
 AsciiTileDir::AsciiTileDir(BlockFile * poFile, uint16 nSegment,
-                           uint32 nBlockSize)
+                           CPL_UNUSED uint32 nBlockSize)
```
or just uint32 /* nBlockSize */

rouault commented on gdal/frmts/pcidsk/sdk/blockdir/asciitiledir.cpp:
```diff
@@ -720,7 +720,7 @@ void AsciiTileDir::WriteDir(void)
     for (size_t iLayer = 0; iLayer < moLayerInfoList.size(); iLayer++)
     {
-        BlockLayerInfo * psLayer = moLayerInfoList[iLayer];
+        psLayer = moLayerInfoList[iLayer];
```
a C++11 range-based loop would even be better

rouault commented:
> "Maybe you guys should think about using the -i option so that we get all the errors."

You're more than welcome to improve the CI.

A few remarks & questions
- This is a quite massive PR. Will be hard to review.
- Is all code added potentially triggered by the GDAL PCIDSK driver ?
- The test suite (autotest/gdrivers/pcidsk.py) could probably be enhanced to trigger the new code paths on sample files (please keep them as small as possible. A few KB max if possible)
- Prior to merging, I'd like to see this code being stress-tested by oss-fuzz (https://github.com/google/oss-fuzz), run locally on your side for a few days. Outline of the procedure (I haven't tested it fully. you may need to do some tweaks):
   * in your fork, apply the below patch so that a dedicated fuzzer for the PCIDSK driver is built
   * also add in autotest/gdrivers/data/pcidsk/ new test cases that will trigger the new code paths
   * commit and push (possibly in a dedicated ossfuzz branch)
   * ``git clone https://github.com/google/oss-fuzz``
   * ``cd oss-fuzz``
   * edit projects/gdal/Dockerfile and modify the "git clone --depth 1 https://github.com/OSGeo/gdal" gdal line to point to your clone and branch
   * ``python infra/helper.py build_image gdal``
   * ``python infra/helper.py build_fuzzers --sanitizer address gdal``
   * ``python infra/helper.py run_fuzzer gdal pcidsk_fuzzer``
   * wait for a few days if it spots bugs, fix, retry the procedure from the stage python infra/helper.py build_image gdal
   * retry with the undefined behaviour santizer: 
   * ``python infra/helper.py build_fuzzers --sanitizer undefined gdal``
   * ``python infra/helper.py run_fuzzer gdal pcidsk_fuzzer``

```
diff --git a/gdal/fuzzers/build_google_oss_fuzzers.sh b/gdal/fuzzers/build_google_oss_fuzzers.sh
index 3b445b4..8d8f229 100755
--- a/gdal/fuzzers/build_google_oss_fuzzers.sh
+++ b/gdal/fuzzers/build_google_oss_fuzzers.sh
@@ -92,7 +92,7 @@ build_ogr_specialized_fuzzer fgb RegisterOGRFlatGeobuf "/vsimem/test.fgb" "/vsim
 build_fuzzer cad_fuzzer $(dirname $0)/ogr_fuzzer.cpp -DREGISTER_FUNC=RegisterOGRCAD
 build_fuzzer rec_fuzzer $(dirname $0)/ogr_fuzzer.cpp -DREGISTER_FUNC=RegisterOGRREC -DUSE_FILESYSTEM -DEXTENSION="\"rec\""
 
-formats="GTiff HFA"
+formats="GTiff HFA PCIDSK"
 for format in $formats; do
     fuzzerName=$(echo $format | tr '[:upper:]' '[:lower:]')
     build_gdal_specialized_fuzzer $fuzzerName "GDALRegister_$format" "/vsimem/test" "/vsimem/test"
```

pciphelippeneveu commented:
Hi,

**This is a quite massive PR. Will be hard to review.**

This should no longer be the case in the future. Internally at PCI, they only use some part of this code. They have returned to using our older version of the code which supports things like adding and removing channels. PCI also added JPEG2000 compression, but have decided to keep these changes private for now. PCI also plans to create a new vector segment to fix the 4GB limit and make it much faster.

**Is all code added potentially triggered by the GDAL PCIDSK driver ?**

Of course not, I did run the new version against our many automated tests and fixed all the errors on our side. I have no clue how your testing works. I cannot even get the automated tests to run on my system. I get the following errors:

E               Failed: data/avc/testpoly.e00
/build/gdal/autotest/ogr/ogr_avc.py:156: Failed

E       AssertionError: assert (-1 != -1)
E        +  where -1 = <built-in method find of str object at 0x7fc8ec20c780>('INFO')
E        +    where <built-in method find of str object at 0x7fc8ec20c780> = '\nERROR ret code = -6'.find
/build/gdal/autotest/ogr/ogr_basic_test.py:405: AssertionError

**The test suite (autotest/gdrivers/pcidsk.py) could probably be enhanced to trigger the new code paths on sample files (please keep them as small as possible. A few KB max if possible)**

Can you tell me how can I manually invoke this pcidsk.py test suite and any other PCIDSK tests? Can you also explain me how I can pass the new file creation option? The PCIDSK format expects: "TILED<size> TILEV1" or "TILED<size> TILEV2". GDAL seems to have separated the options into groups, would the following work?

    tst = gdaltest.GDALTest('PCIDSK', 'png/rgba16.png', 2, 2042,
                            options=['INTERLEAVING=TILED **TILEV1**', 'TILESIZE=32'])

**Prior to merging, I'd like to see this code being stress-tested by oss-fuzz (https://github.com/google/oss-fuzz), run locally on your side for a few days. Outline of the procedure (I haven't tested it fully. you may need to do some tweaks):**

I don't think this will work if I cannot even get the tests running.

Regards,

Phelippe

rouault commented:
> This should no longer be the case in the future

I didn't really understand from your explanations what will change in the future.

> **Is all code added potentially triggered by the GDAL PCIDSK driver ?**
> 
> Of course not

There's maybe a misunderstanding. I didn't mean that the existing tests would trigger the new code paths, but if with the appropriate test files, they could. If some of the code added is never going to be triggered by the GDAL or OGR PCIDSK drivers, then it should be removed, or at the very least not compiled in. GDAL is a big enough beast to allow dead code in it :-)

> I have no clue how your testing works

You could have a look at the scripts run by the CI workers (see in .github/workflows). As they are Docker based, you can also reproduce most of them locally (but do that preferably in a dedicated git worktree)
But basically:
- ./configure --with-python
-  make
- cd apps && make test_ogrsf
- source ./scripts/setdevenv.sh
- pip install -r ../autotest/requirements.txt 
- pytest ../autotest/gdrivers/pcidsk.py ../autotest/ogr/ogr_pcidsk.py
- I believe you'll need to also run `test_ogrsf -driver PCIDSK` as the PR seems to introduces a crash in it

> Can you also explain me how I can pass the new file creation option? 

Perhaps as values TILED_V1 and TILED_V2 for INTERLEAVING ? See frmts/pcidsk/pcidskdataset2.cpp

> I don't think this will work if I cannot even get the tests running.

That's a bit orthogonal, but yes that would be more reasonable to have your GDAL dev env fully functional first before tackling ossfuzz.

pciphelippeneveu commented:
> There's maybe a misunderstanding. I didn't mean that the existing tests would trigger the new code paths, but if with the appropriate test files, they could. If some of the code added is never going to be triggered by the GDAL or OGR PCIDSK drivers, then it should be removed, or at the very least not compiled in. GDAL is a big enough beast to allow dead code in it :-)

There could certainly be some segments that are not used like BreakpointPCT, BreakpointLUT, some math model segments, but I have no clue what GDAL covers.

So if I understand correctly, if I copy some small sample files in autotest/gdrivers/data/pcidsk/ they will be automatically picked up and used in the testing?

> You could have a look at the scripts run by the CI workers (see in .github/workflows). As they are Docker based, you can also reproduce most of them locally (but do that preferably in a dedicated git worktree)

Thank you that was quite helpful. I will take a look at the CI workers.

> Perhaps as values TILED_V1 and TILED_V2 for INTERLEAVING ? See frmts/pcidsk/pcidskdataset2.cpp

Thank you I will try that.

rouault commented:
> There could certainly be some segments that are not used like BreakpointPCT, BreakpointLUT, some math model segments, but I have no clue what GDAL covers.

The answer lies in looking at how pcidskdataset2.cpp (and ogrpcidsklayer.cpp for vector support) uses the SDK. From what I can see it reads band data, metadata, color table, RPC, overviews. Basic write support

> So if I understand correctly, if I copy some small sample files in autotest/gdrivers/data/pcidsk/ they will be automatically picked up and used in the testing?

No, you'll have to add test cases in autotest/gdrivers/pcidsk.py to exercise them (for fuzzing purposes, yes just adding in autotest/gdrivers/data/pcidsk will be enough)

rouault commented on gdal/frmts/pcidsk/sdk/blockdir/asciitilelayer.cpp:
```diff
@@ -0,0 +1,197 @@
+{
+    uint32 nTileCount = GetTileCount();
+
+    size_t nSize = nTileCount * 20;
```
overflow ?

rouault commented on gdal/frmts/pcidsk/sdk/blockdir/asciitilelayer.cpp:
```diff
@@ -0,0 +1,197 @@
+
+    size_t nSize = nTileCount * 20;
+
+    uint8 * pabyTileList = (uint8 *) malloc(nSize);
```
malloc failure?

rouault commented on gdal/frmts/pcidsk/sdk/building.dox:
```diff
@@ -0,0 +1,71 @@
+/*!
```
not sure if that file applies for the SDK inserted in GDAL

rouault commented on gdal/frmts/pcidsk/sdk/core/cpcidskfile.cpp:
```diff
@@ -41,29 +42,50 @@
 #include "core/clinksegment.h"
 
+#ifdef PCIDSK_IMP_PARAM_SUPPORT
+#define __INT16DEF__
```
defines starting with double underscore are reserved in C/C++

rouault commented on gdal/frmts/pcidsk/sdk/core/pcidsk_scanint.cpp:
```diff
@@ -0,0 +1,393 @@
+namespace PCIDSK
+{
+
+int16 ganCharTo1[256] =
```
those tables should be const

rouault commented on gdal/frmts/pcidsk/sdk/segment/vecsegdataindex.cpp:
```diff
@@ -133,22 +133,12 @@ const std::vector<uint32> *VecSegDataIndex::GetIndex()
-            throw PCIDSKException("Out of memory allocating block_index(%u): %s",
-                                  block_count, ex.what());
-        }
+        block_index.resize( block_count );
```
the safety checks from GDAL have been lost here

rouault commented on gdal/frmts/pcidsk/sdk/segment/metadatasegment_p.cpp:
```diff
@@ -102,10 +103,7 @@ void MetadataSegment::Load()
-
-    // data_size < 1024 will throw an exception in SetSize()
-    seg_data.SetSize( data_size < 1024 ? -1 : (int) (data_size - 1024) );
+    seg_data.SetSize( (int) (data_size - 1024) );
```
GDAL safety check lost

rouault commented on gdal/frmts/pcidsk/sdk/segment/cpcidsksegment.cpp:
```diff
@@ -106,26 +105,17 @@ void CPCIDSKSegment::LoadSegmentPointer( const char *segment_pointer )
+    segment_type = static_cast<eSegType>(atoi(segptr.Get(1,3)));
+    if (EQUAL(SegmentTypeName(segment_type), "UNKNOWN"))
+        segment_type = SEG_UNKNOWN;
+    data_offset = (atouint64(segptr.Get(12,11))-1) * 512;
```
Integer overflow checks lost

rouault commented on gdal/frmts/pcidsk/sdk/segment/cpcidskpolymodel.cpp:
```diff
@@ -0,0 +1,317 @@
+        return;
+    }
+
+    assert(data_size - 1024 == 7 * 512);
```
could this assert fail on corrupted files ? If so it should be changed to a runtime check

rouault reviewed (commented):
In a number of occasions, my questions are really questions due to the lack of context & intimate knowledge of the code. For a number of patterns that I've spotted, I've given up spotting all similar occurrences. I let it to you to port the fixes to all relevant similar cases.

pciphelippeneveu commented:
@rouault

I've fixed all failures. I added new tests that covers both tile versions (135K x 2). I've fixed the issues you have raised in your code reviews. The only thing I haven't done is run the PCIDSK format against the oss-fuzz tool.

Note that I did run all of PCI Geomatics' automated tests and everything is passing on our side as well.

I don't think the PCIDSK format is that widely used to begin with.

Let me know if you want me to do anything more for this PR.

Happy Holidays!

Phelippe Neveu

rouault commented:
> The only thing I haven't done is run the PCIDSK format against the oss-fuzz tool.

I'd appreciate if you could give that a try. All that new code has strong potential to trigger new issues, and the overhead of dealing with them after being integrated in master is quite a burden for me.

pciphelippeneveu commented:
> I'd appreciate if you could give that a try. All that new code has strong potential to trigger new issues, and the overhead of dealing with them after being integrated in master is quite a burden for me.

I will give it a try and let you know if I run into issues.

pciphelippeneveu commented:
> * in your fork, apply the below patch so that a dedicated fuzzer for the PCIDSK driver is built
> 
> * also add in autotest/gdrivers/data/pcidsk/ new sample files that will trigger the new code paths

Do I need to add the pcidsk test files to any configuration file?

rouault commented:
> Do I need to add the pcidsk test files to any configuration file?

Ah, yes to make the fuzzer more efficient, you could edit fuzzers/build_seed_corpus.sh to create a seed corpus from the sample files for the pcidsk_fuzzer. Look for example at how hfa_fuzzer_seed_corpus.zip is built (you'll need only to zip files in autotest/gdrivers/data/pcidsk)

rouault commented:
One thing to try would also to build & run the sanitizers for 32 bit builds.
e.g: python infra/helper.py build_fuzzers --sanitizer address --architecture i386 gdal

rouault commented:
> 1 issue that I haven't fixed because it would have required changes to the cpl::VSICurlHandle::Read() method.

What was the issue ? Fixes for other parts of GDAL are welcome (in another PR)

pciphelippeneveu commented:
> What was the issue ? Fixes for other parts of GDAL are welcome (in another PR)

gdal/port/cpl_vsil_curl.cpp: 1838
iterOffset - nOffsetToDownload

pciphelippeneveu commented:
@rouault

I'm done with running the different oss-fuzz modes on the PCIDSK library. So I believe that I'm finally done with this PR. It certainly was a longer process than anticipated. Let me know if you want me to do anything else. Note that all the tests are also passing on our side.

At the end, I cloned 3 oss-fuzz branches for the different modes. I then ran each mode in parallel using a bash script that runs the pcidsk_fuzzer in a loop using a different log file for every run so that I could inspect the crashes after letting it run overnight. The only failures left are all related to the cpl::VSICurlHandle::Read() method.

> Fixes for other parts of GDAL are welcome (in another PR)

Is there a way to run the test with one of the "build/out/gdal/crash-<key>" output file?

Cheers!

Phelippe Neveu

rouault commented:
> Is there a way to run the test with one of the "build/out/gdal/crash-" output file?

yes, see https://google.github.io/oss-fuzz/advanced-topics/reproducing/#reproducing-bugs

pciphelippeneveu commented:
Anything else to do for this PR?

rouault commented:
Just squashed & merged. Thanks for your work

pciphelippeneveu commented:
Thank you, going to try and fix the cpl::VSICurlHandle::Read() crash in another PR.
